apps/ejabberd/src/mod_mam_utils.erl

```diff
@@ -177,11 +177,9 @@ microseconds_to_datetime(MicroSeconds) when is_integer(MicroSeconds) ->
 -spec generate_message_id() -> integer().
 generate_message_id() ->
     {ok, NodeId} = ejabberd_node_id:node_id(),
-    %% Unique enough
-    {T1, T2, T3} = p1_time_compat:timestamp(),
-    ImprovedTS = {T1, T2,
-                  (T3 div 1000) * 1000 + p1_time_compat:unique_integer([positive]) rem 1000 },
-    encode_compact_uuid(now_to_microseconds(ImprovedTS), NodeId).
+    CandidateStamp = p1_time_compat:system_time(micro_seconds),
```

Comment:
I'd use `p1_time_compat:os_system_time(micro_seconds)` as with time warp disabled by default, some environments could move Erlang system time into the future and never become correct until VM restart.